File: simple-cqrs/Repository.cs
using System;
using LanguageExt;
using static LanguageExt.Prelude;

namespace SimpleCQRS
{
   public interface IRepository<T> where T : Aggregate
   {
      Either<string, Unit> Save(T aggregate);
      Either<string, (T, Seq<Event>)> Load(Guid rootId, Func<Seq<Event>, (T, Seq<Event>)> factory);
   }

   public static class RepositoryExt
   {
      public static Either<string, (T, Seq<Event>)> Load<T>(this IRepository<T> repo, ICommand cmd) where T : Aggregate =>
         repo.Load(cmd.RootId, x => Aggregate.Cons<T>(x));

      public static Either<string, (T, Seq<Event>)> LoadExpectedVersion<T>(this IRepository<T> repo, ICommand cmd, int expectedVersion) where T : Aggregate =>
         repo.Load(cmd.RootId, x => Aggregate.Cons<T>(x))
            .Bind(x => x.Item1.Info.Version == expectedVersion
               ? x
               : Left<string, (T, Seq<Event>)>("aggregate version is different"));
   }

   public sealed class Repository<T> : IRepository<T> where T : Aggregate
   {
      private readonly IEventStore _storage;

      public Repository(IEventStore storage)
      {
         _storage = storage;
      }

      public Either<string, (T, Seq<Event>)> Load(Guid rootId, Func<Seq<Event>, (T, Seq<Event>)> factory) =>
         _storage.GetEvents(rootId)
            .Map(factory);

      public Either<string, Unit> Save(T aggregate) =>
         aggregate
            .NewEvents
            .Map(x => new Event(x.AggregateInfo, x.Timestamp, x.Owner, x.EventData))
            .Map(x => _storage.AddEvent(x))
            .Sequence()
            .Match(
               Right: _ => _storage.Save(),
               Left: err => err);
   }
}

File: Fescq/Repository.fs
namespace Fescq

open System
open Fescq.Aggregate

type IRepository<'t> =

   abstract member Save : Agg<'t> * Event list -> Result<Agg<'t>, string>

   /// input: aggregate ID * factory
   /// output: aggregate * history
   abstract member Load : Guid * (Event list -> Result<Agg<'t>, string>) -> Result<(Agg<'t> * Event list), string>


type Repository<'t> (storage:IEventStore) =

   interface IRepository<'t> with

      member x.Load (aggregateId:Guid, factory:(Event list -> Result<Agg<'t>, string>)) =
         aggregateId
         |> storage.GetEvents
         |> Result.bind (fun events ->
               events
               |> factory
               |> Result.bind (fun agg -> Ok (agg, events)
               )
            )

      member x.Save (aggregate:Agg<'t>, unpersisted:Event list) =
         if aggregate.History.Length > 0 
            && unpersisted.Length > 0
            && (List.last unpersisted) = (List.last aggregate.History) then
            
            // bail at the first error
            // https://stackoverflow.com/a/26890974/571637
            unpersisted
            |> List.unfold (fun events ->
                  match events with 
                  | head :: tail ->
                     match storage.AddEvent head with
                     | Ok _ -> Some (Ok (), tail)
                     | Error msg -> Some (Error msg, []) 
                  | [] -> None)

            |> List.last 
            |> Result.bind storage.Save
            |> Result.bind (fun _ -> aggregate |> Ok)
         else
            Error "unpersisted events must be part of aggregate history"

(*

public static class RepositoryExt
{
   public static Either<string, (T, Seq<Event>)> Load<T>(this IRepository<T> repo, ICommand cmd) where T : Aggregate =>
      repo.Load(cmd.RootId, x => Aggregate.Cons<T>(x));

   public static Either<string, (T, Seq<Event>)> LoadExpectedVersion<T>(this IRepository<T> repo, ICommand cmd, int expectedVersion) where T : Aggregate =>
      repo.Load(cmd.RootId, x => Aggregate.Cons<T>(x))
         .Bind(x => x.Item1.Info.Version == expectedVersion
            ? x
            : Left<string, (T, Seq<Event>)>("aggregate version is different"));
}


*)